protoc-gen-gogo/generator/generator.go

```diff
@@ -3209,6 +3217,106 @@ func (g *Generator) generateCommonMethods(mc *msgCtx) {
 	g.P("var xxx_messageInfo_", mc.goName, " ", g.Pkg["proto"], ".InternalMessageInfo")
 }
 
+func (g *Generator) cloneGenericHeader(typeName, returnName string) {
+	g.P("func (m *", typeName, ") Clone() ", returnName, "{")
+	g.In()
+	g.P("if m == nil {")
+	g.In()
+	g.P("return nil")
+	g.endBlock()
+	g.P("cloned := new(", typeName, ")")
+	g.P("*cloned = *m")
+	g.P()
+}
+
+func (g *Generator) cloneGenericFooter() {
+	g.P("return cloned")
+	g.endBlock()
+}
+
+func (g *Generator) ifNotNilHeader(name string) {
+	g.P("if m.", name, "!= nil {")
+	g.In()
+}
+
+func (g *Generator) endBlock() {
+	g.Out()
+	g.P("}")
+}
+
+func (g *Generator) generateClone(mc *msgCtx, topLevelFields []topLevelField, mapFields map[string]*GoMapDescriptor) {
+	g.cloneGenericHeader(mc.goName, fmt.Sprintf("*%s", mc.goName))
```

Comment:
For upstreaming: there should be a "generic" `Clone() proto.Message` (or `proto.CloneableMessage`) such that you can have a `proto.CloneableMessage` interface that the `proto.Clone()` implementation tests for